### PR TITLE
chore: Use Go-based implementation as embedded extension to Python SDK

### PR DESCRIPTION
**Pull request by pyalex**
**What this PR does / why we need it**:

Converts Go-based implementation of get online features functionality into library which will be embedded into Python code and called directly from Python SDK (w/o GRPC in the middle). This allows us to also use Python functions as callbacks to call some implementations (like online/offline store connectors) from Go feature server.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: Makefile
Comment (woop) [inline Makefile:80]: Should these `go get` statements be in a `install-go-ci-dependencies` command?

File: go/internal/feast/onlinestore.go
Comment (woop) [inline go/internal/feast/onlinestore.go:18]: Why are these being capitalized? Our convention thus far has been to use lower case unless we are referring to classes/objects.

Comment (pyalex) [reply go/internal/feast/onlinestore.go:18]: some autoreplace, will fix

File: go/internal/feast/featurestore.go
Comment (woop) [inline go/internal/feast/featurestore.go:246]: Is there a way that we can make `kind` here a type instead of `interface{}`. Haven't looked to closely at it though.

File: go/embedded/online_features.go
Comment (woop) [inline go/embedded/online_features.go:50]: Would you mind dropping some comments in this function? I think it would be pretty informative for contributors to understand how it functions and how it will be used.

File: go/embedded/online_features.go
Comment (woop) [inline go/embedded/online_features.go:67]: I guess this has nothing to do with your PR, but do you think it may be more natural for these internal functions to be `fs.GetOnlinFeaturesByFeatureService()` and `fs.GetOnlineFeaturesByFeatureRefs()`?

File: go/internal/feast/featurestore.go
Comment (woop) [inline go/internal/feast/featurestore.go:33]: Should this be called `FeatureRefStrings` or something to be more explicit?

File: go/utils/typeconversion.go
Comment (woop) [inline go/utils/typeconversion.go:13]: So arrow boolean maps to Feast bytes (as opposed to https://github.com/feast-dev/feast/blob/bb8274147888ba70ee50282653f2aec566672911/protos/feast/types/Value.proto#L34)

File: go/internal/feast/featurestore.go
Comment (kevjumba) [inline go/internal/feast/featurestore.go:44]: I feel like this should be capitalized to follow convention for datatypes `FeatureAndView`

File: sdk/python/feast/embedded_go/online_features_service.py
Comment (woop) [inline sdk/python/feast/embedded_go/online_features_service.py:23]: Should this be an exception? What happens if the import fails? I see `Datatable` being used below which will cause an exception there.

File: go/internal/feast/featurestore.go
Comment (pyalex) [reply go/internal/feast/featurestore.go:44]: The convention is to capitalize names that should be exposed. This is internal type

File: go/utils/typeconversion.go
Comment (felixwang9817) [inline go/utils/typeconversion.go:83]: is this a duplicate of `ProtoTypeToArrowType` above?

File: go/internal/feast/featurestore.go
Comment (felixwang9817) [reply go/internal/feast/featurestore.go:33]: how about just `FeatureRefs`?

File: go/internal/feast/featurestore.go
Comment (felixwang9817) [inline go/internal/feast/featurestore.go:49]: nit: can we have a comment for this struct?

File: go/internal/feast/featurestore.go
Comment (felixwang9817) [inline go/internal/feast/featurestore.go:55]: nit: line 54 above is a stray comment I believe

File: go/internal/feast/featurestore.go
Comment (felixwang9817) [inline go/internal/feast/featurestore.go:37]: nit: can we have comments for new structs describing their purpose?

File: go/internal/feast/featurestore.go
Comment (felixwang9817) [reply go/internal/feast/featurestore.go:44]: this name is somewhat confusing at first to me - I don't feel strongly, but maybe something like `featureViewAndRefs` would be better?

File: go/internal/feast/featurestore.go
Comment (felixwang9817) [reply go/internal/feast/featurestore.go:246]: where is `ExtractFeatureRefs` being used?

**Review by felixwang9817 — COMMENTED** [review]
@pyalex some preliminary comments; will continue reviewing

File: sdk/python/tests/integration/online_store/test_universal_online.py
Comment (kevjumba) [inline sdk/python/tests/integration/online_store/test_universal_online.py:896]: Why do we have to delete these tests? They're the only tests currently working that tests the go feature server.

File: sdk/python/feast/embedded_go/online_features_service.py
Comment (kevjumba) [inline sdk/python/feast/embedded_go/online_features_service.py:12]: I think this import should be from `feast.protos.feast.serving.ServingService_pb2.GetOnlineFeaturesResponse.FeatureVector`, not from the tests

File: go/utils/typeconversion.go
Comment (kevjumba) [reply go/utils/typeconversion.go:13]: Why not use binary or fixed sized binary for bytes?

File: go/embedded/online_features.go
Comment (pyalex) [reply go/embedded/online_features.go:50]: Added some comments (see fun definition)

File: go/internal/feast/featurestore.go
Comment (pyalex) [reply go/internal/feast/featurestore.go:44]: Added

File: go/internal/feast/featurestore.go
Comment (pyalex) [reply go/internal/feast/featurestore.go:63]: added

File: go/internal/feast/featurestore.go
Comment (pyalex) [reply go/internal/feast/featurestore.go:68]: fixed

File: sdk/python/tests/integration/online_store/test_universal_online.py
Comment (pyalex) [reply sdk/python/tests/integration/online_store/test_universal_online.py:896]: What's the purpose of this tests? It doesn't check anything go server related. This is just duplicate of `test_online_store_cleanup`

Comment (pyalex) [reply sdk/python/tests/integration/online_store/test_universal_online.py:896]: `test_go_server_life_cycle ` was testing process management. There's no process anymore

File: sdk/python/feast/embedded_go/online_features_service.py
Comment (pyalex) [reply sdk/python/feast/embedded_go/online_features_service.py]: Great catch. Fixed

File: sdk/python/tests/integration/online_store/test_universal_online.py
Comment (kevjumba) [reply sdk/python/tests/integration/online_store/test_universal_online.py:896]: Oh right I didn't actually check, I was under the impression that we only had 2 go online tests but we had 3, I was mainly worried about online_retrieval_with_go_server but that is still there!

File: go/utils/typeconversion.go
Comment (pyalex) [reply go/utils/typeconversion.go]: Fixed

File: go/internal/feast/featurestore.go
Comment (pyalex) [reply go/internal/feast/featurestore.go]: Updated

**Review by felixwang9817 — COMMENTED** [review]
more comments, will continue looking

File: sdk/python/tests/integration/online_store/test_universal_online.py
Comment (felixwang9817) [inline sdk/python/tests/integration/online_store/test_universal_online.py:868]: why was this change made?

File: sdk/python/tests/conftest.py
Comment (felixwang9817) [inline sdk/python/tests/conftest.py:97]: why are we removing the `goserver` mark? I feel like it will still be useful from a developer perspective to be able to run the specific tests that rely on the go server

if you feel strongly that you want to remove `goserver`, you should also remove `goserver` from Makefile

File: sdk/python/feast/feature_store.py
Comment (felixwang9817) [inline sdk/python/feast/feature_store.py:94]: just curious, why do we need to wrap this with `TYPE_CHECKING`?

File: go/internal/feast/featurestore_test.go
Comment (achals) [inline go/internal/feast/featurestore_test.go:52]: This test is currently skipped; should we unskip it if it's working locally now?

File: sdk/python/feast/embedded_go/online_features_service.py
Comment (achals) [inline sdk/python/feast/embedded_go/online_features_service.py:66]: Some comments here would be nice

File: sdk/python/tests/integration/online_store/test_universal_online.py
Comment (pyalex) [reply sdk/python/tests/integration/online_store/test_universal_online.py:868]: Entities `customer_id` and `driver_id` are not used in this request. And Go implementation explicitly returns error when some unexpected join keys were passed. In comparison Python implementation allow to pass any entities that exists in the registry, but also throw an error when entity name is not recognized.

File: sdk/python/tests/conftest.py
Comment (pyalex) [reply sdk/python/tests/conftest.py:97]: Yeah, the idea was not to have go dedicated tests at all, and instead have dedicate test environment with flag. Currently the only blocker is support of ODFV on Go side. So I will rollback this for now, until tests are merged.

File: sdk/python/feast/feature_store.py
Comment (pyalex) [reply sdk/python/feast/feature_store.py:94]: This enables lazy importing of `EmbeddedOnlineFeatureServer`. In case if somebody would need to run feast locally (set up from GH, eg) and fix something small w/o need to build go library. Not very strong case, but still could be useful.

File: go/cmd/server/server.go
Comment (pyalex) [reply go/cmd/server/server.go:46]: This loop populate response with entities. Next loop populate with features.

File: go/internal/feast/featurestore_test.go
Comment (pyalex) [reply go/internal/feast/featurestore_test.go:52]: This doesn't work locally yet, since requires registry, which apparently should be set up externally.

File: sdk/python/feast/embedded_go/online_features_service.py
Comment (pyalex) [reply sdk/python/feast/embedded_go/online_features_service.py]: Added

File: go/embedded/online_features.go
Comment (pyalex) [reply go/embedded/online_features.go]: Removed

File: go/internal/feast/featurestore_test.go
Comment (pyalex) [reply go/internal/feast/featurestore_test.go]: Added comment. It has columnar structure: 3 columns of features and 1 column with join key values

File: Makefile
Comment (pyalex) [reply Makefile:130]: Added ToDo

**Comment by pyalex** [comment]
@achals @felixwang9817 Addressed your comments.

File: go/cmd/server/server.go
Comment (felixwang9817) [reply go/cmd/server/server.go:46]: got it - I'm fine with leaving this as is (no need for comment), I would just note that it was super confusing bc the proto definition is
```
message GetOnlineFeaturesResponseMetadata {
    FeatureList feature_names = 1;
}
```
which doesn't really make it seem like there's any entities involved

**Review by felixwang9817 — COMMENTED** [review]
I'm comfortable with merging in this PR now; I'll let @achals have the final say

thanks @pyalex for addressing my comments!

File: go/cmd/server/server.go
Comment (pyalex) [reply go/cmd/server/server.go:46]: yeah, I understand. But this implementation just keeps compatibility with Python implementation that does the same as of this moment.

**Review by achals — COMMENTED** [review]
looks mostly fine, but the biggest piece is the featurestore.go file which I don't think I have confidence that I've reviewed thoroughly. I'll take a pass over that tomorrow morning.

File: Makefile
Comment (pyalex) [reply Makefile]: this is embarrassing) fixed

File: go/internal/feast/featurestore.go
Comment (achals) [inline go/internal/feast/featurestore.go:624]: nit, can we have a different module name instead of `utils`? Seems like it'll become a kitchen sink..

**Review by achals — APPROVED** [review]
/lgtm

File: go/internal/feast/featurestore.go
Comment (pyalex) [reply go/internal/feast/featurestore.go]: renamed into `types` since it's type utils

**Pull request merged**
